src/MiddlewareListener.php

```diff
@@ -96,6 +117,42 @@ public function onDispatch(MvcEvent $event)
     }
 
     /**
+     * Create a middleware pipe from the array spec given.
+     *
+     * @param ContainerInterface $serviceLocator
+     * @param ResponseInterface $responsePrototype
+     * @param array $middlewaresToBePiped
+     * @return MiddlewarePipe
+     * @throws \InvalidArgumentException
+     * @throws \Zend\Mvc\Exception\MiddlewareNotCallableException
+     */
+    private function createPipeFromSpec(
+        ContainerInterface $serviceLocator,
+        ResponseInterface $responsePrototype,
+        array $middlewaresToBePiped
+    ) {
+        $pipe = new MiddlewarePipe();
+        $pipe->setResponsePrototype($responsePrototype);
+        foreach ($middlewaresToBePiped as $middlewareToBePiped) {
+            if (null === $middlewareToBePiped) {
+                throw new \InvalidArgumentException('Middleware name cannot be null');
+            }
+
+            $middlewareName = is_string($middlewareToBePiped) ? $middlewareToBePiped : get_class($middlewareToBePiped);
+
+            if (is_string($middlewareToBePiped) && $serviceLocator->has($middlewareToBePiped)) {
+                $middlewareToBePiped = $serviceLocator->get($middlewareToBePiped);
+            }
+            if (! is_callable($middlewareToBePiped)) {
```

Comment:
Once you make those changes, you'll also need to add tests for piping http-interop middleware.